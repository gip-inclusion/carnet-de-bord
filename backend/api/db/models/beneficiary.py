from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field

from api.db.models.notebook import Notebook


class Beneficiary(BaseModel):
    id: UUID
    email: str | None
    lastname: str
    firstname: str
    caf_number: str | None
    pe_number: str | None
    postal_code: str | None
    city: str | None
    address1: str | None
    address2: str | None
    mobile_number: str | None
    date_of_birth: date
    place_of_birth: str | None
    deployment_id: UUID
    created_at: datetime
    updated_at: datetime
    internal_id: str | None
    notebook: Notebook | None
    # BRSA users may not have an account
    # (account is created on the  first login attempt)
    account_id: UUID | None


class BeneficiaryImport(BaseModel):
    si_id: str = Field(..., title="Identifiant du SI interne")
    firstname: str = Field(..., title="Prénom")
    lastname: str = Field(..., title="Nom")
    date_of_birth: date = Field(..., title="Date de naissance")
    place_of_birth: str | None = Field(None, title="Lieu de naissance")
    phone_number: str | None = Field(None, title="Numéro de téléphone")
    email: str | None = Field(None, title="Email")
    address1: str | None = Field(None, title="Adresse")
    address2: str | None = Field(None, title="Complément d'adresse")
    postal_code: str | None = Field(None, title="Code postal")
    city: str | None = Field(None, title="Ville")
    work_situation: str | None = Field(None, title="Situation professionnelle")
    caf_number: str | None = Field(None, title="Numéro CAF/MSA")
    pe_number: str | None = Field(None, title="Identifiant Pole Emploi")
    right_rsa: str | None = Field(None, title="Droits RSA")
    right_are: bool | None = Field(None, title="Droits ARE")
    right_ass: bool | None = Field(None, title="Droits ASS")
    right_bonus: bool | None = Field(None, title="Droits Bonus")
    right_rqth: bool | None = Field(None, title="Droits RQTH")
    geographical_area: str | None = Field(None, title="Zone de mobilité")
    rome_code_description: str | None = Field(None, title="Emploi recherché")
    education_level: str | None = Field(None, title="Niveau de formation")
    structure_name: str | None = Field(None, title="Structure d'accompagnement")
    advisor_email: str | None = Field(None, title="Accompagnateur référent")
