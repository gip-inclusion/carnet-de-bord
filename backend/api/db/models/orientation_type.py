from strenum import StrEnum


class OrientationType(StrEnum):
    pro = "pro"
    social = "social"
    sociopro = "sociopro"
