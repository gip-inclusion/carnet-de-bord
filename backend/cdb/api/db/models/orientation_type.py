from strenum import StrEnum


class OrientationType(StrEnum):
    pro = "pro"
    social = "social"
    sociopro = "sociopro"

    def get_label(self):
        if self.value == "pro":
            return "Professionnel"
        elif self.value == "social":
            return "Social"
        elif self.value == "sociopro":
            return "Socio-professionnel"
