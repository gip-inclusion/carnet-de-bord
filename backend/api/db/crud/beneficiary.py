from datetime import date
from typing import List
from uuid import UUID

from asyncpg import Record
from asyncpg.connection import Connection
from asyncpg.exceptions import UniqueViolationError

from api.db.crud.notebook import (
    NOTEBOOK_BASE_FIELDS,
    NOTEBOOK_BASE_JOINS,
    parse_notebook_from_record,
)
from api.db.models.beneficiary import Beneficiary, BeneficiaryImport

BENEFICIARY_BASE_QUERY = (
    """SELECT b.*, acc.id as account_id,"""
    + NOTEBOOK_BASE_FIELDS
    + """
    FROM public.beneficiary b
    LEFT JOIN notebook n ON n.beneficiary_id = b.id
    """
    + NOTEBOOK_BASE_JOINS
    + """
    LEFT JOIN public.account acc
    ON acc.beneficiary_id = b.id
"""
)


# see https://betakuang.medium.com/why-postgresqls-on-conflict-cannot-find-my-partial-unique-index-552327b85e1
firstname_lastname_date_of_birth_unique_idx = (
    "(LOWER(trim(firstname)), LOWER(trim(lastname)), date_of_birth, deployment_id)"
)


async def select_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    return await connection.fetch(
        f"""
SELECT (firstname, lastname, date_of_birth, internal_id, deployment_id, id)
FROM beneficiary
WHERE (firstname = $1 AND lastname = $2 AND date_of_birth = $3)
OR (internal_id = $4 AND deployment_id = $5)
        """,
        beneficiary.firstname,
        beneficiary.lastname,
        beneficiary.date_of_birth,
        beneficiary.si_id,
        deployment_id,
    )


async def update_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    await connection.fetchrow(
        f"""
UPDATE beneficiary SET mobile_number = $1
where internal_id = $2
returning id
        """,
        beneficiary.phone_number,
        beneficiary.si_id,
    )


async def insert_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    await connection.fetchrow(
        f"""
INSERT INTO BENEFICIARY (firstname, lastname, internal_id, date_of_birth, deployment_id, mobile_number)
values ($1, $2, $3, $4, $5, $6)
returning id
        """,
        beneficiary.firstname,
        beneficiary.lastname,
        beneficiary.si_id,
        beneficiary.date_of_birth,
        deployment_id,
        beneficiary.phone_number,
    )


async def import_beneficiary(
    connection: Connection,
    beneficiary: BeneficiaryImport,
    deployment_id,
):
    result = select_beneficiary(connection, beneficiary, deployment_id)
    existing_rows = await result
    if len(existing_rows) == 0:
        print("inserting new user")
        await insert_beneficiary(connection, beneficiary, deployment_id)
    else:
        print(existing_rows[0][0][4])
        if (
            len(existing_rows) == 1
            and existing_rows[0][0][0] == beneficiary.firstname
            and existing_rows[0][0][1] == beneficiary.lastname
            and existing_rows[0][0][2] == beneficiary.date_of_birth
            and existing_rows[0][0][3] == beneficiary.si_id
        ):
            # Doesn't work somehow
            #     and existing_rows[0][0][4] == deployment_id
            print("updating existing user")
            await update_beneficiary(connection, beneficiary, deployment_id)
        else:
            print("data problem")


async def get_beneficiary_with_query(
    connection: Connection, query: str, *args
) -> Beneficiary | None:

    async with connection.transaction():

        beneficiary = None

        beneficiary_records: List[Record] = await connection.fetch(
            BENEFICIARY_BASE_QUERY + query,
            *args,
        )

        for beneficiary_record in beneficiary_records:

            if beneficiary is None:
                beneficiary = Beneficiary.parse_obj(beneficiary_record)

            if beneficiary_record["n_id"] is not None:
                beneficiary.notebook = await parse_notebook_from_record(
                    beneficiary_record,
                    notebook=beneficiary.notebook,
                )

        return beneficiary


async def get_beneficiary_from_personal_information(
    connection: Connection, firstname: str, lastname: str, birth_date: date
) -> Beneficiary | None:

    return await get_beneficiary_with_query(
        connection,
        # We use LOWER(BTRIM(field)) here to match the index so postgres
        # can use the index rather than scanning the whole table
        "WHERE LOWER(BTRIM(b.firstname)) = LOWER($1) AND "
        "LOWER(BTRIM(b.lastname)) = LOWER($2) AND "
        "b.date_of_birth = $3",
        firstname,
        lastname,
        birth_date,
    )


async def get_beneficiary_by_id(
    connection: Connection, beneficiary_id: UUID
) -> Beneficiary | None:

    return await get_beneficiary_with_query(
        connection, "WHERE b.id = $1", beneficiary_id
    )
