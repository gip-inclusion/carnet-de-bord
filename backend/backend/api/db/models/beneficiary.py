from datetime import date, datetime
from uuid import UUID

from pandas.core.series import Series
from pydantic import BaseModel, EmailStr, Field, ValidationError, validator

from backend.api.db.models.csv import CsvFieldError
from backend.api.db.models.nir import nir_format
from backend.api.db.models.notebook import Notebook
from backend.api.db.models.validator import (
    date_validator,
    parse_bool_validator,
    phone_validator,
    postal_code_validator,
)


class BeneficiaryStructure(BaseModel):
    structure_id: UUID
    structure_name: str
    beneficiary_status: str


class Beneficiary(BaseModel):
    id: UUID
    email: str | None
    lastname: str
    firstname: str
    caf_number: str | None
    pe_number: str | None
    postal_code: str | None
    city: str | None
    address1: str | None
    address2: str | None
    mobile_number: str | None
    date_of_birth: date
    place_of_birth: str | None
    deployment_id: UUID
    created_at: datetime
    updated_at: datetime
    internal_id: str | None
    notebook: Notebook | None
    # BRSA users may not have an account
    # (account is created on the first login attempt)
    account_id: UUID | None
    nir: str | None
    pe_unique_import_id: str | None


def snake_to_camel(field):
    parts = field.split("_")
    return "".join(parts[0:1] + [part.title() for part in parts[1:]])


def is_same_name(firstname1, firstname2, lastname1, lastname2):
    return (
        firstname1.lower().strip() == firstname2.lower().strip()
        and lastname1.lower().strip() == lastname2.lower().strip()
    )


class BeneficiaryImport(BaseModel):
    internal_id: str = Field(..., alias="Identifiant dans le SI*")
    firstname: str = Field(..., alias="Prénom*")
    lastname: str = Field(..., alias="Nom*")
    date_of_birth: date = Field(..., alias="Date de naissance*")
    place_of_birth: str | None = Field(None, alias="Lieu de naissance")
    mobile_number: str | None = Field(None, alias="Téléphone")
    email: EmailStr | None = Field(None, alias="Email")
    address1: str | None = Field(None, alias="Adresse")
    address2: str | None = Field(None, alias="Adresse (complément)")
    postal_code: str | None = Field(None, alias="Code postal")
    city: str | None = Field(None, alias="Ville")
    work_situation: str | None = Field(None, alias="Situation")
    caf_number: str | None = Field(None, alias="Numéro allocataire CAF/MSA")
    pe_number: str | None = Field(None, alias="Identifiant Pôle emploi")
    right_rsa: str | None = Field(None, alias="Droits RSA")
    right_are: bool | None = Field(None, alias="Droits ARE")
    right_ass: bool | None = Field(None, alias="Droits ASS")
    right_bonus: bool | None = Field(None, alias="Prime d'activité")
    right_rqth: bool | None = Field(None, alias="RQTH")
    geographical_area: str | None = Field(None, alias="Zone de mobilité")
    rome_code_description: str | None = Field(
        None, alias="Emploi recherché (code ROME)"
    )
    education_level: str | None = Field(None, alias="Niveau de formation")
    structure_name: str | None = Field(None, alias="Structure")
    advisor_email: EmailStr | None = Field(None, alias="Accompagnateurs")
    nir: str | None = Field(None, alias="NIR")

    class Config:
        anystr_strip_whitespace = True
        allow_population_by_field_name = True

    _phone_validator = phone_validator("mobile_number")
    _postal_code_validator = postal_code_validator("postal_code")
    _parse_bool_validator = parse_bool_validator(
        "right_are", "right_ass", "right_bonus", "right_rqth", pre=True
    )
    _date_validator = date_validator("date_of_birth", pre=True)

    @validator("right_rsa")
    def parse_right_rsa(cls, right_rsa):
        if right_rsa and right_rsa not in [
            "rsa_droit_ouvert_et_suspendu",
            "rsa_droit_ouvert_versable",
            "rsa_droit_ouvert_versement_suspendu",
        ]:
            raise ValueError("value unknown")
        return right_rsa

    @validator("geographical_area")
    def parse_geographical_area(cls, geographical_area):
        if geographical_area and geographical_area not in [
            "none",
            "less_10",
            "between_10_20",
            "between_20_30",
            "plus_30",
        ]:
            raise ValueError("value unknown")
        return geographical_area

    @validator("work_situation")
    def parse_work_situation(cls, work_situation):
        if work_situation and work_situation not in [
            "recherche_emploi",
            "recherche_formation",
            "recherche_alternance",
            "recherche_stage",
            "emploi",
            "formation",
            "stage",
            "alternance",
            "service_civique",
            "iae",
            "interim",
            "construction_projet",
            "projet_entrepreneurial",
            "entrepreneur",
            "etudiant",
            "scolarisé",
            "enretraite",
            "maladie",
            "invalidite",
            "conge_parental",
            "au_foyer",
            "autre",
            "cdi_temps_plein",
            "cdi_temps_partiel",
            "cdd_temps_plein",
            "cdd_temps_partiel",
            "intermittent",
        ]:
            raise ValueError("value unknown")
        return work_situation

    @validator("education_level")
    def parse_education_level(cls, education_level):
        if education_level and education_level not in [
            "AFS",
            "C12",
            "C3A",
            "CFG",
            "CP4",
            "NV5",
            "NV4",
            "NV3",
            "NV2",
            "NV1",
        ]:
            raise ValueError("value unknown")
        return education_level

    @validator("nir")
    def parse_nir(cls, nir: str):
        validation_error = nir_format(nir)
        if validation_error:
            raise ValueError(validation_error)
        return nir

    # TODO Should be possible to use decorator to mark field for beneficiary
    def get_beneficiary_editable_keys(self) -> list[str]:
        BENEFICIARY_EDITABLE_FIELDS = [
            "place_of_birth",
            "mobile_number",
            "address1",
            "address2",
            "postal_code",
            "city",
            "caf_number",
            "pe_number",
            "email",
            "nir",
        ]
        return self._filter_updatable_field_keys(BENEFICIARY_EDITABLE_FIELDS)

    # TODO Should be possible to use decorator to mark field for notebook
    def get_notebook_editable_keys(self) -> list[str]:
        NOTEBOOK_EDITABLE_FIELDS = [
            "right_rsa",
            "right_rqth",
            "right_are",
            "right_ass",
            "right_bonus",
            "work_situation",
            "education_level",
            "geographical_area",
        ]
        return self._filter_updatable_field_keys(NOTEBOOK_EDITABLE_FIELDS)

    def _filter_updatable_field_keys(self, allowed_fields) -> list[str]:
        filled_fields = self._get_filled_field_keys()
        keys_to_update = [key for key in filled_fields if key in allowed_fields]
        # Special case for address where we need to clean address2 if address1 is provided but not address2
        if "address1" in keys_to_update and "address2" not in keys_to_update:
            keys_to_update.append("address2")

        return keys_to_update

    def get_values_for_keys(self, keys: list[str]) -> list[object]:
        return [self.dict().get(key) for key in keys]

    def _get_filled_field_keys(self):
        return [
            fieldname for (fieldname, value) in self.dict().items() if value is not None
        ]


class BeneficiaryCsvRowResponse(BaseModel):
    row: dict | None = None
    data: BeneficiaryImport | None = None
    valid: bool
    update: bool = False
    errors: list[CsvFieldError] | None = None


def map_csv_row(row: Series) -> BeneficiaryCsvRowResponse:
    try:
        data = BeneficiaryImport.parse_obj(row)
        return BeneficiaryCsvRowResponse(data=data, valid=True)

    except ValidationError as error:
        return BeneficiaryCsvRowResponse(
            row=row.to_dict(),
            valid=False,
            errors=[
                CsvFieldError(key="".join(str(err["loc"][0])), error=err["msg"])
                for err in error.errors()
            ],
        )
