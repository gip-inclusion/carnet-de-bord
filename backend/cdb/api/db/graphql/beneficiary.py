import json
from datetime import date
from typing import List
from uuid import UUID

from gql import gql
from gql.client import AsyncClientSession
from graphql import DocumentNode
from pydantic import BaseModel, Field

from cdb.api.db.models.validator import (
    date_validator,
)
from cdb.caf_msa.parse_infos_foyer_rsa import CdbBeneficiaryInfos
from cdb.cdb_csv.json_encoder import CustomEncoder


class BeneficiaryRsaInfos(BaseModel):
    id: UUID
    right_rsa: str | None = Field(alias="rightRsa")
    rsa_closure_reason: str | None = Field(alias="rsaClosureReason")
    rsa_closure_date: date | None = Field(alias="rsaClosureDate")
    rsa_suspension_reason: str | None = Field(alias="rsaSuspensionReason")
    is_homeless: bool | None = Field(alias="isHomeless")
    subject_to_right_and_duty: bool | None = Field(alias="subjectToRightAndDuty")
    caf_number: str = Field(alias="cafNumber")

    _date_validator = date_validator("rsa_closure_date", pre=True)

    class Config:
        anystr_strip_whitespace = True
        allow_population_by_field_name = True


async def get_beneficiary_by_nir(
    gql_session: AsyncClientSession, nir: str
) -> BeneficiaryRsaInfos | None:
    beneficiary_object: dict[str, List[dict]] = await gql_session.execute(
        beneficiary_by_nir_request(),
        variable_values={"nir": nir},
    )
    return parse_beneficiaries(beneficiary_object)


async def update_beneficiary(
    gql_session: AsyncClientSession,
    id: UUID,
    beneficiary_infos: CdbBeneficiaryInfos,
    sha: str,
    external_data: dict,
) -> BeneficiaryRsaInfos | None:
    result: dict[str, dict] = await gql_session.execute(
        update_beneficiary_mutation(),
        variable_values={
            "id": str(id),
            "rsaInfos": beneficiary_infos.dict(),
            "hash": sha,
            # Hack @lionelb:
            # Using a custom encoder on gql save a string containing
            # the stringify version on the json
            # that's why we use a encode / decode strategy
            # to transform non serializable object into string
            # so final dict can be saved into jsonb field
            "externalData": json.loads(json.dumps(external_data, cls=CustomEncoder)),
        },
    )
    beneficiary = result.get("beneficiary")
    if beneficiary:
        return parse_beneficiary(beneficiary)


def parse_beneficiaries(data: dict[str, List[dict]]) -> BeneficiaryRsaInfos | None:
    beneficiaries = data.get("beneficiaries")
    if isinstance(beneficiaries, List) and len(beneficiaries) == 1:
        return parse_beneficiary(beneficiaries[0])


def parse_beneficiary(data: dict) -> BeneficiaryRsaInfos:
    return BeneficiaryRsaInfos.parse_obj(data)


def beneficiary_by_nir_request() -> DocumentNode:
    return gql(
        """
query beneficiary($nir: String!) {
  beneficiaries:get_beneficiaries_from_nir(args:{search_nir: $nir}){
    id
    rightRsa
    rsaClosureReason
    rsaClosureDate
    rsaSuspensionReason
    isHomeless
    subjectToRightAndDuty
    cafNumber
    lastname
  }
}
"""
    )


def update_beneficiary_mutation() -> DocumentNode:
    return gql(
        """
mutation update_beneficiary(
    $id: uuid!,
    $rsaInfos: beneficiary_set_input!,
    $externalData: jsonb!,
    $hash: String!
) {
  beneficiary: update_beneficiary_by_pk(pk_columns: {id: $id}, _set: $rsaInfos ) {
    id
    rightRsa
    rsaClosureReason
    rsaClosureDate
    rsaSuspensionReason
    isHomeless
    subjectToRightAndDuty
    cafNumber
    lastname
  }
  insert_external_data_info_one(object: {
    beneficiary_id: $id
    externalData: {
      data: {
        data: $externalData
        source: cafmsa
        hash: $hash
      }
    }
  }){
    external_data_id
  }
}
        """
    )
