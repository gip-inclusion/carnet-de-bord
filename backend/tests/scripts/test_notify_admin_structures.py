from datetime import date
from unittest import mock

from asyncpg.connection import Connection

from cdb.api.db.models.beneficiary import BeneficiaryWithAdminStructureEmail
from cdb.api.db.models.orientation_type import OrientationType
from cdb.scripts.notify_admin_structures.notify_admin_structures import (
    notify_admin_structures,
)


@mock.patch("cdb.api.core.emails.send_mail")
@mock.patch(
    "cdb.scripts.notify_admin_structures.notify_admin_structures.get_beneficiaries_without_referent"
)
class TestNotifyAdminStructure:
    async def test_notify_admin_structures_when_no_beneficiaries_without_referent(
        self,
        mock_get_beneficiaries_without_referent: mock.Mock,
        mock_send_email: mock.Mock,
        db_connection: Connection,
    ):
        mock_get_beneficiaries_without_referent.return_value = []

        await notify_admin_structures(db_connection)

        mock_send_email.assert_not_called()

    async def test_notify_admin_structures_when_one_beneficiary_without_referent_with_one_admin_structure(
        self,
        mock_get_beneficiaries_without_referent: mock.Mock,
        mock_send_email: mock.Mock,
        db_connection: Connection,
        snapshot,
    ):

        mock_get_beneficiaries_without_referent.return_value = [
            BeneficiaryWithAdminStructureEmail(
                firstname="Jay",
                lastname="Pasderéférent",
                date_of_birth=date(1981, 12, 20),
                admin_structure_email="admin@structure.cdb",
                structure_name="structure 123",
                orientation=None,
            )
        ]

        await notify_admin_structures(db_connection)

        mock_send_email.assert_called_once()
        assert (
            mock_send_email.call_args_list[0].kwargs["subject"]
            == "Nouveaux bénéficiaires sans référent"
        )
        assert mock_send_email.call_args_list[0].kwargs["to"] == "admin@structure.cdb"
        assert snapshot == mock_send_email.call_args_list[0].kwargs["message"]

    async def test_notify_admin_structures_when_one_beneficiary_without_referent_with_two_admin_structures(
        self,
        mock_get_beneficiaries_without_referent: mock.Mock,
        mock_send_email: mock.Mock,
        db_connection: Connection,
        snapshot,
    ):
        mock_get_beneficiaries_without_referent.return_value = [
            BeneficiaryWithAdminStructureEmail(
                firstname="Jay",
                lastname="Pasderéférent",
                date_of_birth=date(1981, 12, 20),
                admin_structure_email="admin@structure.cdb",
                structure_name="structure 123",
                orientation=None,
            ),
            BeneficiaryWithAdminStructureEmail(
                firstname="Jay",
                lastname="Pasderéférent",
                date_of_birth=date(1981, 12, 20),
                admin_structure_email="second_admin@structure.cdb",
                structure_name="structure 123",
                orientation=OrientationType.pro,
            ),
        ]

        await notify_admin_structures(db_connection)

        assert mock_send_email.call_count == 2
        assert (
            mock_send_email.call_args_list[0].kwargs["subject"]
            == "Nouveaux bénéficiaires sans référent"
        )
        assert mock_send_email.call_args_list[0].kwargs["to"] == "admin@structure.cdb"
        assert (
            mock_send_email.call_args_list[1].kwargs["subject"]
            == "Nouveaux bénéficiaires sans référent"
        )
        assert (
            mock_send_email.call_args_list[1].kwargs["to"]
            == "second_admin@structure.cdb"
        )
        assert snapshot == [
            mock_send_email.call_args_list[0].kwargs["message"],
            mock_send_email.call_args_list[1].kwargs["message"],
        ]

    async def test_notify_admin_structures_when_two_beneficiaries_without_referent_with_one_admin_structure_and_two_structures(
        self,
        mock_get_beneficiaries_without_referent: mock.Mock,
        mock_send_email: mock.Mock,
        db_connection: Connection,
        snapshot,
    ):
        mock_get_beneficiaries_without_referent.return_value = [
            BeneficiaryWithAdminStructureEmail(
                firstname="Jay",
                lastname="Pasderéférent",
                date_of_birth=date(1981, 12, 20),
                admin_structure_email="admin@structure.cdb",
                structure_name="structure 123",
                orientation=OrientationType.social,
            ),
            BeneficiaryWithAdminStructureEmail(
                firstname="Jean",
                lastname="Naipanonplu",
                date_of_birth=date(1980, 5, 7),
                admin_structure_email="admin@structure.cdb",
                structure_name="structure ABC",
                orientation=OrientationType.sociopro,
            ),
        ]

        await notify_admin_structures(db_connection)

        assert mock_send_email.call_count == 1
        assert (
            mock_send_email.call_args_list[0].kwargs["subject"]
            == "Nouveaux bénéficiaires sans référent"
        )
        assert mock_send_email.call_args_list[0].kwargs["to"] == "admin@structure.cdb"
        assert snapshot == mock_send_email.call_args_list[0].kwargs["message"]
