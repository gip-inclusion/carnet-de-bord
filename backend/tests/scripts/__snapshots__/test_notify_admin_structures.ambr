# name: TestNotifyAdminStructure.test_notify_admin_structures_when_one_beneficiary_without_referent_with_one_admin_structure
  '''
  <!DOCTYPE html>
  <html lang="fr">
  	<head>
  		<meta charset="UTF-8" />
  		<meta http-equiv="X-UA-Compatible" content="IE=edge" />
  		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
  		<title>Bénéficiaires en attente de la désignation d'un référent</title>
  	</head>
  	<body>
  		<h1>Bénéficiaires en attente de la désignation d'un référent</h1>
  		<p>
  			Bonjour,
  		</p>
  
  		
  			<p>
  				Ci-dessous, la liste de(s) bénéficiaire(s) orienté(s) vers la structure
  				structure 123 pour lesquels un référent doit être désigné.
  			</p>
  
  			<table>
  				<tr>
  					<th>Bénéficiaire</th>
  					<th>Date de naissance</th>
  					<th>Accompagnement</th>
  				</tr>
  				
  				<tr>
  					<td>Jay Pasderéférent</td>
  					<td>20/12/1981</td>
  					<td>-</td>
  				</tr>
  				
  			</table>
  		
  
  			<p>
  			Pour accéder à Carnet de bord, veuillez cliquer sur le lien
  			ci-dessous&nbsp;:
  		</p>
  		<p style="padding-left: 20%">
  			<a
  				rel="nofollow"
  				href="http://localhost:3001"
  				style="
  					background-color: #6a20ae;
  					font-size: 14px;
  					font-family: Helvetica, Arial, sans-serif;
  					font-weight: bold;
  					text-decoration: none;
  					padding: 8px 10px;
  					color: #ffffff;
  					border-radius: 5px;
  					display: inline-block;
  				"
  			>
  				<span>Accédez à Carnet de bord</span>
  			</a>
  		</p>
  		<p>L'équipe Carnet de bord</p>
  	</body>
  </html>
  '''
# ---
# name: TestNotifyAdminStructure.test_notify_admin_structures_when_one_beneficiary_without_referent_with_two_admin_structures
  list([
    '''
      <!DOCTYPE html>
      <html lang="fr">
      	<head>
      		<meta charset="UTF-8" />
      		<meta http-equiv="X-UA-Compatible" content="IE=edge" />
      		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
      		<title>Bénéficiaires en attente de la désignation d'un référent</title>
      	</head>
      	<body>
      		<h1>Bénéficiaires en attente de la désignation d'un référent</h1>
      		<p>
      			Bonjour,
      		</p>
      
      		
      			<p>
      				Ci-dessous, la liste de(s) bénéficiaire(s) orienté(s) vers la structure
      				structure 123 pour lesquels un référent doit être désigné.
      			</p>
      
      			<table>
      				<tr>
      					<th>Bénéficiaire</th>
      					<th>Date de naissance</th>
      					<th>Accompagnement</th>
      				</tr>
      				
      				<tr>
      					<td>Jay Pasderéférent</td>
      					<td>20/12/1981</td>
      					<td>-</td>
      				</tr>
      				
      			</table>
      		
      
      			<p>
      			Pour accéder à Carnet de bord, veuillez cliquer sur le lien
      			ci-dessous&nbsp;:
      		</p>
      		<p style="padding-left: 20%">
      			<a
      				rel="nofollow"
      				href="http://localhost:3001"
      				style="
      					background-color: #6a20ae;
      					font-size: 14px;
      					font-family: Helvetica, Arial, sans-serif;
      					font-weight: bold;
      					text-decoration: none;
      					padding: 8px 10px;
      					color: #ffffff;
      					border-radius: 5px;
      					display: inline-block;
      				"
      			>
      				<span>Accédez à Carnet de bord</span>
      			</a>
      		</p>
      		<p>L'équipe Carnet de bord</p>
      	</body>
      </html>
    ''',
    '''
      <!DOCTYPE html>
      <html lang="fr">
      	<head>
      		<meta charset="UTF-8" />
      		<meta http-equiv="X-UA-Compatible" content="IE=edge" />
      		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
      		<title>Bénéficiaires en attente de la désignation d'un référent</title>
      	</head>
      	<body>
      		<h1>Bénéficiaires en attente de la désignation d'un référent</h1>
      		<p>
      			Bonjour,
      		</p>
      
      		
      			<p>
      				Ci-dessous, la liste de(s) bénéficiaire(s) orienté(s) vers la structure
      				structure 123 pour lesquels un référent doit être désigné.
      			</p>
      
      			<table>
      				<tr>
      					<th>Bénéficiaire</th>
      					<th>Date de naissance</th>
      					<th>Accompagnement</th>
      				</tr>
      				
      				<tr>
      					<td>Jay Pasderéférent</td>
      					<td>20/12/1981</td>
      					<td>Professionnel</td>
      				</tr>
      				
      			</table>
      		
      
      			<p>
      			Pour accéder à Carnet de bord, veuillez cliquer sur le lien
      			ci-dessous&nbsp;:
      		</p>
      		<p style="padding-left: 20%">
      			<a
      				rel="nofollow"
      				href="http://localhost:3001"
      				style="
      					background-color: #6a20ae;
      					font-size: 14px;
      					font-family: Helvetica, Arial, sans-serif;
      					font-weight: bold;
      					text-decoration: none;
      					padding: 8px 10px;
      					color: #ffffff;
      					border-radius: 5px;
      					display: inline-block;
      				"
      			>
      				<span>Accédez à Carnet de bord</span>
      			</a>
      		</p>
      		<p>L'équipe Carnet de bord</p>
      	</body>
      </html>
    ''',
  ])
# ---
# name: TestNotifyAdminStructure.test_notify_admin_structures_when_two_beneficiaries_without_referent_with_one_admin_structure_and_two_structures
  '''
  <!DOCTYPE html>
  <html lang="fr">
  	<head>
  		<meta charset="UTF-8" />
  		<meta http-equiv="X-UA-Compatible" content="IE=edge" />
  		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
  		<title>Bénéficiaires en attente de la désignation d'un référent</title>
  	</head>
  	<body>
  		<h1>Bénéficiaires en attente de la désignation d'un référent</h1>
  		<p>
  			Bonjour,
  		</p>
  
  		
  			<p>
  				Ci-dessous, la liste de(s) bénéficiaire(s) orienté(s) vers la structure
  				structure 123 pour lesquels un référent doit être désigné.
  			</p>
  
  			<table>
  				<tr>
  					<th>Bénéficiaire</th>
  					<th>Date de naissance</th>
  					<th>Accompagnement</th>
  				</tr>
  				
  				<tr>
  					<td>Jay Pasderéférent</td>
  					<td>20/12/1981</td>
  					<td>Social</td>
  				</tr>
  				
  			</table>
  		
  			<p>
  				Ci-dessous, la liste de(s) bénéficiaire(s) orienté(s) vers la structure
  				structure ABC pour lesquels un référent doit être désigné.
  			</p>
  
  			<table>
  				<tr>
  					<th>Bénéficiaire</th>
  					<th>Date de naissance</th>
  					<th>Accompagnement</th>
  				</tr>
  				
  				<tr>
  					<td>Jean Naipanonplu</td>
  					<td>07/05/1980</td>
  					<td>Socio-professionnel</td>
  				</tr>
  				
  			</table>
  		
  
  			<p>
  			Pour accéder à Carnet de bord, veuillez cliquer sur le lien
  			ci-dessous&nbsp;:
  		</p>
  		<p style="padding-left: 20%">
  			<a
  				rel="nofollow"
  				href="http://localhost:3001"
  				style="
  					background-color: #6a20ae;
  					font-size: 14px;
  					font-family: Helvetica, Arial, sans-serif;
  					font-weight: bold;
  					text-decoration: none;
  					padding: 8px 10px;
  					color: #ffffff;
  					border-radius: 5px;
  					display: inline-block;
  				"
  			>
  				<span>Accédez à Carnet de bord</span>
  			</a>
  		</p>
  		<p>L'équipe Carnet de bord</p>
  	</body>
  </html>
  '''
# ---
